test/integration/whois_server_integration_test.rb

```diff
@@ -0,0 +1,225 @@
+require 'test_helper'
+require 'simpleidn'
+require_relative '../../lib/whois_server'
+require_relative '../../lib/whois_server_core'
+require_relative '../../app/models/whois_record'
+require_relative '../../app/validators/unicode_validator'
+
+class WhoisServerIntegrationTest < Minitest::Test
+  include WhoisServer
+
+  def setup
+    super
+    @sent_data = []
+    @connection_closed = false
+    @connection_closed_after_writing = false
+    @logger_output = StringIO.new
+    @logger = Logger.new(@logger_output)
+    ENV['WHOIS_ENV'] = 'test'
+  end
+
+  def test_receive_data_with_invalid_data
+    stubs(:invalid_data?).returns(true)
+    stubs(:invalid_encoding_msg).returns('Invalid encoding')
+    stubs(:connection).returns(nil)
+
+    receive_data('invalid\xFF')
+
+    assert_equal 1, @sent_data.length
+    assert_includes @sent_data.first, 'Invalid encoding'
+    assert @connection_closed_after_writing
+  end
+
+  def test_receive_data_with_no_ip
+    stubs(:extract_ip).returns(nil)
+    stubs(:connection).returns(nil)
+
+    receive_data('example.ee')
+
+    assert_equal 0, @sent_data.length
+  end
+
+  def test_receive_data_full_flow

```

Comment:
What exactly is this test checking? You’ve mocked so many methods that I can’t tell what we’re actually testing here.